src/Core/src/Platform/Windows/MauiWebView.cs

```diff
@@ -1,15 +1,39 @@
 ﻿using System;
 using System.Diagnostics;
 using System.Text.RegularExpressions;
+using Microsoft.Maui.ApplicationModel;
 using Microsoft.UI.Xaml.Controls;
+using Windows.ApplicationModel;
 
 namespace Microsoft.Maui.Platform
 {
 	public class MauiWebView : WebView2, IWebViewDelegate
 	{
+		public MauiWebView()
+		{
+			NavigationStarting += (sender, args) =>
+			{
+				// Auto map local virtual app dir host, e.g. if navigating back to local site from a link to an external site
+				if (args?.Uri?.ToLowerInvariant().StartsWith(LocalScheme.TrimEnd('/').ToLowerInvariant()) == true)
+				{
+					CoreWebView2.SetVirtualHostNameToFolderMapping(
+						LocalHostName,
+						ApplicationPath,
+						Web.WebView2.Core.CoreWebView2HostResourceAccessKind.Allow);
+				}
+				// Auto unmap local virtual app dir host if navigating to any other potentially unsafe domain
+				else
+				{
+					CoreWebView2.ClearVirtualHostNameToFolderMapping(LocalHostName);
+				}
+			};
+		}
+
 		WebView2? _internalWebView;
 
-		const string LocalScheme = "ms-appx-web:///";
+		// Arbitrary local host name for virtual folder mapping
+		const string LocalHostName = "appdir";
```

Comment:
I wonder if you wanted to add some more "Defense In Depth", you could have the local hostname be _random_, meaning that even if for some reason the hostname mapping was still present, no one would know what it was.

This is similar to what Chromium-based browsers (Chrome, Edge) do for extensions:

<img width="352" alt="screenshot of randomized browser extension URL" src="https://user-images.githubusercontent.com/202643/174681133-806d88fe-ced0-4390-b4b9-42730046eb91.png">

You could have the hostname be something like `appdir_[50 random digits]`, and generated when the control loads.

I assume there's no reason for the app to ever care what the hostname is statically, right? If so, this behavior could go very far in mitigating any potential problems.

Reply:
Hmm... I am curious what the default behaviour is on Android & iOS for this. Will run same test as above on those platforms to see...

In terms of whether the hostname needs to be constant, I don't believe it does. Even if for some reason someone needed it to be, they could just add a handler in their own app code anyway (as I have done temporarily to work around this issue), as you can map more than one virtual hostname in the WebView2. If they load an absolute URL the built-in mapping won't even kick in either.

But I'm happy to make these changes if you think it's necessary and will make this airtight.

Reply:
Yeah good question about what do the other webviews do. They probably offer very little protection. Let's skip this idea for now and just stick with what you have.